pkg/executor/sortexec/sort_partition.go

```diff
@@ -0,0 +1,311 @@
+		}
+	})
+
+	sort.Slice(s.savedRows, s.keyColumnsLess)
+	s.isSorted = true
+	return
+}
+
+func (s *sortPartition) spillToDiskImpl() error {
+	s.inDisk = chunk.NewDataInDiskByChunks(s.fieldTypes)
+	s.diskTracker.AttachTo(s.diskTracker)
+	tmpChk := chunk.NewChunkWithCapacity(s.fieldTypes, spillChunkSize)
+
+	rowNum := len(s.savedRows)
+	if rowNum == 0 {
+		return errSpillEmptyChunk
+	}
+
+	for ; s.idx < rowNum; s.idx++ {
+		row := s.savedRows[s.idx]
+		tmpChk.AppendRow(row)
+		if tmpChk.IsFull() {
+			err := s.inDisk.Add(tmpChk)
+			if err != nil {
+				return err
+			}
+			tmpChk.Reset()
+		}
+	}
+
+	// Spill the remaining data in tmpChk.
+	// Do not spill when tmpChk is empty as `Add` function requires a non-empty chunk
+	if tmpChk.NumRows() > 0 {
+		err := s.inDisk.Add(tmpChk)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Release memory as all data have been spilled to disk
+	s.savedRows = nil
+	s.getMemTracker().Consume(-s.totalTrackedMemNum)
+	return nil
+}
+
+// We can only call this function under the protection of `syncLock`.
+func (s *sortPartition) spillToDisk() error {
+	err := s.sort()
+	if err != nil {
+		return err
+	}
+
+	// We should consider spill has been triggered once goroutine get `syncLock`.
+	// However, `s.sort()` will retrigger the spill and lead to dead lock
+	// so we have to set spill flag at here.
+	s.setSpillTriggered()
+	s.helper.setIsSpilling(true)
+	err = s.spillToDiskImpl()
+	s.helper.setIsSpilling(false)
+	s.helper.cond.Broadcast()
+	return err
+}
+
+func (s *sortPartition) actionSpill(helper *spillHelper) *sortPartitionSpillDiskAction {
+	if s.spillAction == nil {
+		s.spillAction = &sortPartitionSpillDiskAction{
+			partition: s,
+			helper:    helper,
+		}
+	}
+	return s.spillAction
+}
+
+func (s *sortPartition) getSortedRowFromMemoryAndAppendToChunk(filledChk *chunk.Chunk) {
+	filledChk.AppendRow(s.savedRows[s.idx])
+	s.advanceIdx()
+}
+
+func (s *sortPartition) getMemTracker() *memory.Tracker {
+	return s.memTracker
+}
+
+func (s *sortPartition) getDiskTracker() *disk.Tracker {
+	return s.diskTracker
+}
+
+func (s *sortPartition) hasEnoughDataToSpill() bool {
+	// Guarantee that each partition size is not too small, to avoid opening too many files.
+	return s.getMemTracker().BytesConsumed() > s.spillLimit
+}
+
+func (s *sortPartition) lessRow(rowI, rowJ chunk.Row) bool {
+	for i, colIdx := range s.keyColumns {
+		cmpFunc := s.keyCmpFuncs[i]
+		if cmpFunc != nil {
+			cmp := cmpFunc(rowI, colIdx, rowJ, colIdx)
+			if s.byItemsDesc[i] {
+				cmp = -cmp
+			}
+			if cmp < 0 {
+				return true
+			} else if cmp > 0 {
+				return false
+			}
+		}
+	}
+	return false
+}
+
+func (s *sortPartition) numRowInMemory() int {
+	return len(s.savedRows)
+}
+
+// keyColumnsLess is the less function for key columns.
+func (s *sortPartition) keyColumnsLess(i, j int) bool {
+	if s.timesOfRowCompare >= signalCheckpointForSort {
+		// Trigger Consume for checking the NeedKill signal
+		s.memTracker.Consume(1)
```

Comment:
If sort_partition.tracker is not attached to global tracker, there seems to be no way to watch the kill signal?

Reply:
> If sort_partition.tracker is not attached to global tracker, there seems to be no way to watch the kill signal?

No, attaching to session is enough.